prov/verbs/src/fi_verbs.c

```diff
@@ -609,6 +612,14 @@ static int fi_ibv_read_params(void)
 			   "Invalid value of mr_max_cached_size\n");
 		return -FI_EINVAL;
 	}
+	if (fi_ibv_get_param_bool("mr_cache_merge_regionss",
+				  "Enable the merging of MR regions for MR "
+				  "caching functionality",
+				  &fi_ibv_gl_data.mr_cache_merge_regions)) {
```

Comment:
I dread seeing new environment variables.  Can this be hidden from the user with the correct behavior chosen for them?  Or maybe change how merging works to fix whatever problem this is trying to address?

Reply:
> Or maybe change how merging works to fix whatever problem this is trying to address?

The merging works fine. But this combines very huge buffers into one region for some MPI ring based algorithms (an user buffer is cut into small ones). `ibv_reg_mr` fails for some reason due to lack of resources in this case.
I tried to limit merging of regions by combining only small regions. This works, but performance is becoming bad in this case.

I'm going to make MR caching enabled for verbs provider by default. The MR merging is one thing that is show stopper to do this. MR caching w/o this mechanism works perfect and I prove the stability of this feature on IB and iWarp.

Reply:
Hmm... my first thought was to add a size limit to any registered region that can be cached, and restrict merged regions from going past that size.  Any idea why the performance was bad limiting merging to small regions only?

Reply:
> Any idea why the performance was bad limiting merging to small regions only?

It seems this depends on how an user operates with memory regions.
For example, the limiting to small regions only (e.x. limit == 16 Kbytes) should outperform in the following case:
- cutting 16 Kbytes buffer into small ones (1024 buffer of 16 bytes each)
- Several times does with this 16 Kbytes buffer the following:
```
iter = 1000;
while (iter--) {
   for (i = 0 ; i  < 1024; i+)+ {
     fi_mr_reg(buf[i], mr);
     fi_send(buf[i], 16);
     fi_close(mr);
   }
}
```
i.e.  registration of each buffer and sending it out and then close MR
This case gives us possibility to re-use the same MR (that is result of merging several regions) multiple times.

Reply:
I'm confused now.  You originally said, "I tried to limit merging of regions by combining only small regions. This works, but performance is becoming bad in this case."  But this is saying that merging small regions was better performance.

In any case, there's definitely overhead for merging.  If the merged region isn't re-used enough times, the cost will be higher than keeping the regions separate.  One thing we can do is:

If the region will be used for local access only (FI_SEND and/or FI_RECV), we should be able to safely register the entire page, rather than just the specified size.